sdk/core/core-rest-pipeline/src/nodeHttpClient.ts

```diff
@@ -243,28 +246,47 @@ class NodeHttpClient implements HttpClient {
   }
 
   private getOrCreateAgent(request: PipelineRequest, isInsecure: boolean): http.Agent {
-    if (!request.disableKeepAlive) {
-      if (isInsecure) {
-        if (!this.httpKeepAliveAgent) {
-          this.httpKeepAliveAgent = new http.Agent({
-            keepAlive: true,
-          });
-        }
+    const disableKeepAlive = request.disableKeepAlive;
 
-        return this.httpKeepAliveAgent;
-      } else {
-        if (!this.httpsKeepAliveAgent) {
-          this.httpsKeepAliveAgent = new https.Agent({
-            keepAlive: true,
-          });
-        }
+    // Handle Insecure requests first
+    if (isInsecure) {
+      if (disableKeepAlive) {
+        // keepAlive:false is the default so we don't need a custom Agent
+        return http.globalAgent;
+      }
 
-        return this.httpsKeepAliveAgent;
+      if (!this.cachedHttpAgent) {
+        // If there is no cached agent create a new one and cache it.
+        this.cachedHttpAgent = new http.Agent({ keepAlive: true });
       }
-    } else if (isInsecure) {
-      return http.globalAgent;
+      return this.cachedHttpAgent;
     } else {
-      return https.globalAgent;
+      if (disableKeepAlive && !request.tlsSettings) {
+        // When there are no tlsSettings and keepAlive is false
+        // we don't need a custom agent
+        return https.globalAgent;
+      }
+
+      // We use the tlsSettings to index cached clients
+      const tlsSettings = request.tlsSettings ?? DEFAULT_TLS_SETTINGS;
+
+      // Get the cached agent or create a new one with the
+      // provided values for keepAlive and tlsSettings
+      let agent = this.cachedHttpsAgents.get(tlsSettings);
+
+      if (agent && agent.options.keepAlive === !disableKeepAlive) {
+        return agent;
+      }
+
+      agent = new https.Agent({
```

Comment:
I wonder if it would be useful to have a `debug` level log here that we're creating a new agent. The only reason I'm thinking about it is if someone provides inline TLS settings (instead of reusing the same object) we'll create a new Agent every time which I think would maybe hurt performance. At least if there is a log about how "no cached TLS Agent exists" or something, they might root cause it faster.